feat: 비밀번호 찾기 프론트 & 백엔드 [ISSUE-408, 465]

aegis1920 opened the pull request:
**이슈 번호**

resolved: #408 
resolved: #465

**작업 내용**

프론트

1. 비밀번호 찾기 페이지 구현
2. 비밀번호 찾기 API 구현
3. 로그인 페이지에 비밀번호 찾기 Link

백엔드

1. MailAuth가 이미 있다면 업데이트하고, 없으면 추가
2. 랜덤 비밀번호는 MailAuthService에 있던 것 사용
3. 인증을 여러 번 눌러도 전송했다고 하는 이슈 해결
4. 랜덤 비밀번호 자리수 증가
5. Mail 클래스를 만들어 좀 더 메일을 작성하기 편하도록 수정
6. 회원가입 submit 실패 시 전부 reset하지 않는 이슈 해결
7. 비밀번호 찾기 및 인증 컨트롤러 구현 

<img width="375" alt="Screen Shot 2020-09-25 at 3 46 06 AM" src="https://user-images.githubusercontent.com/22311557/94192466-f3d6d800-fee9-11ea-9a98-037a28266854.png">
<img width="308" alt="Screen Shot 2020-09-25 at 3 48 24 AM" src="https://user-images.githubusercontent.com/22311557/94192409-e1f53500-fee9-11ea-8d24-527e0e587c58.png">
<img width="654" alt="Screen Shot 2020-09-25 at 3 46 56 AM" src="https://user-images.githubusercontent.com/22311557/94192441-eb7e9d00-fee9-11ea-9a95-1a92203dde17.png">
<img width="662" alt="Screen Shot 2020-09-25 at 3 47 23 AM" src="https://user-images.githubusercontent.com/22311557/94192431-e91c4300-fee9-11ea-8501-9ba167ed4951.png">
<img width="665" alt="Screen Shot 2020-09-25 at 3 49 14 AM" src="https://user-images.githubusercontent.com/22311557/94192252-a8bcc500-fee9-11ea-9628-7d5a28315427.png">

**테스트 작성 여부**

- [X] Test case

aegis1920 committed:
feat: 비밀번호 찾기 인수테스트 작성 및 Mock Controller

aegis1920 committed:
feat: PasswordFindService 클래스 추가 및 테스트 작성, AuthUtils 분리

- existsByNickname, existsByEmail 메서드 추가

aegis1920 committed:
feat: FindController 추가, PasswordFindService 와 FindController 연결, 인수테스트 메일 인증 Mocking

- Mail 도메인 추가
- 인증 암호를 100_000_000 ~ 900_000_000 으로 변경

aegis1920 committed:
feat: 비밀번호 찾기 페이지 및 API 구현

- 로그인 페이지에 비밀번호 찾기 Link
- 로그인 페이지 validate 활성화

yelimkim98 commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/dto/PasswordFindRequest.java:
```diff
@@ -0,0 +1,18 @@
+@Getter
+@EqualsAndHashCode
+@ToString
+public class PasswordFindRequest {
+    private String nickname;
```
여기에 공백 한 줄 넣어주면 더 좋을것같아요!

yelimkim98 reviewed (approved):
오오 수고하셨습니다!! 👍

jamie9504 commented:
랜덤 비밀번호를 전달해주는 것보다, 회원가입때처럼 인증번호를 받아서 비밀번호를 변경하는 방식이 더 좋다고 생각했었는데

이 부분 회의 때 같이 얘기해보면 좋을 것 같아요 :)

aegis1920 committed:
merge develop branch

aegis1920 committed:
Merge branch 'feature/#465' of https://github.com/aegis1920/2020-legeno-around-here into feature/#408

aegis1920 committed:
fix: 필요 없는 파일 삭제

aegis1920 committed:
feat: 비밀번호 찾기 인수테스트 및 컨트롤러 작성, DTO 작성

- WebSecurityConfig 에 "users/me/password" 추가
- 비밀번호 찾기 컨트롤러 작성
- 비밀번호를 바꾸는 컨트롤러 권한 유무로 분리
- DTO도 권한 유무로 분리
-

aegis1920 committed:
feat: 서비스 기능 구현 및 서비스 테스트 작성

- 비밀번호를 바꾸는 서비스를 권한 유무로 분리
- 메일 저장 시 이미 있으면 바꾸고, 없으면 추가

aegis1920 committed:
feat: 비밀번호 찾기 페이지 구현

- 인증을 여러 번 눌러도 전송했다고 하는 이슈 해결

aegis1920 committed:
feat: 비밀번호 찾기 API 구현

aegis1920 committed:
fix: 회원가입 시 인증 여러 번 눌렀을 때도 전송했다고 말하는 이슈, sumit 실패 시 전부 reset하지 않는 이슈 해결

aegis1920 committed:
Merge branch 'develop' of https://github.com/woowacourse-teams/2020-legeno-around-here into feature/#408

aegis1920 commented:
회의 후, 인증 번호를 받아서 비밀번호를 변경할 수 있도록 구현했습니다 :) 조언 감사합니닷!

jamie9504 commented on back-end/legeno-around-here/src/main/java/wooteco/team/ittabi/legenoaroundhere/controller/UserController.java:
```diff
@@ -102,10 +103,20 @@
     }
 
+    @PutMapping(USERS_PATH + ME_PATH + "/password-auth")
+    public ResponseEntity<Void> changeMyPasswordWithAuth(
+        @RequestBody UserPasswordWithAuthUpdateRequest userPasswordWithAuthUpdateRequest) {
+        userService.changeMyPasswordWithAuth(userPasswordWithAuthUpdateRequest);
+
+        return ResponseEntity
+            .noContent()
+            .build();
+    }
+
     @PutMapping(USERS_PATH + ME_PATH + "/password")
     public ResponseEntity<Void> changeMyPassword(
         @RequestBody UserPasswordUpdateRequest userPasswordUpdateRequest) {
-        userService.changeMyPassword(userPasswordUpdateRequest);
+        userService.changeMyPasswordWithoutAuth(userPasswordUpdateRequest);
```
클래스랑 서비스랑 명명을 비슷하게 가도 될 것 같아요.

aegis1920 replied:
오 이 부분 놓치고 있었네요! 감사합니다 :)

jamie9504 commented on back-end/legeno-around-here/src/test/java/wooteco/team/ittabi/legenoaroundhere/utils/constants/UserConstants.java:
```diff
@@ -21,9 +21,11 @@
     public static final String TEST_NEW_USER_NICKNAME = "n_userName";
     public static final String TEST_NEW_USER_PASSWORD = "n_userPassword";
 
+
```
```suggestion
```

:>

jamie9504 commented on front-end/legeno-around-here/src/components/api/API.js:
```diff
@@ -64,6 +67,7 @@ export const createPost = async (postData, accessToken, history) => {
   } catch (error) {
     if (await redirectLoginWhenUnauthorized(error, history)) return;
 
+    console.log(error);
```
console.log를 지워봅시다!

jamie9504 commented on front-end/legeno-around-here/src/components/api/API.js:
```diff
@@ -586,7 +633,7 @@ export const findOthersProfileById = async ({ accessToken, userId, history }) =>
 
       if (error.response && error.response.status === 404) {
-        alert('존재하지 않는 회원입니다.');
+        alert("존재하지 않는 회원입니다.");
```
"가 들어간걸 보아, 프리티어가 잘 적용이 안된 것 같네용 ?!

aegis1920 replied:
수정했습니다~

jamie9504 commented on front-end/legeno-around-here/src/components/pages/LoginPage.js:
```diff
@@ -125,6 +126,11 @@ const LoginForm = ({ history }) => {
             </Grid>
+            <Grid item>
+              <LinkWithoutStyle to='/find/password' variant='body2'>
+                {'비밀번호를 까먹으셨나요?'}
```
까먹으셨나요보단 잊어버리셨나요가 더 나은 것 같아요..

aegis1920 replied:
수정했어요! 두 개와 비교해보니 '잊어버리셨나요'가 더 맞는 것 같습니다 :)

jamie9504 commented on front-end/legeno-around-here/src/components/pages/find/FindPasswordPage.js:
```diff
@@ -0,0 +1,267 @@
+  const classes = useStyles();
+  const PASSWORD_MIN_LENGTH = 8;
+  const PASSWORD_MAX_LENGTH = 16;
+  const EMAIL_REGEX = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
```
JOIN PAGE에도 있었던 것 같은데 같이 써도 될 듯 하네용!

aegis1920 replied:
곧 데모라 이 부분은 추후에 Utils로 통합하도록 하겠습니다!

jamie9504 reviewed (approved):
고생하셨어요!

데모 끝나고나면 추후에, 비밀번호 찾기는 Provider - LOCAL 제한을 걸어둬도 좋겠네요! 코멘트 조금 남기고 어프로브합니다~!

aegis1920 committed:
chore: 공백, 함수명 등 피드백 반영
